app/models/user.rb

```diff
@@ -4,10 +4,40 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
 
+  # Virtual attribute for authenticating by either username or email
+  # This is in addition to a real persisted field like 'username'
+  attr_accessor :login
+
+  validates_format_of :username, with: /^[a-zA-Z0-9_\.]*$/, multiline: true
+  validates :username,
+  presence: true,
+  uniqueness: {
+    case_sensitive: false
+  }
+
   has_one :homepage, dependent: :destroy
   after_create -> { self.create_homepage }
 
   has_many :lists
 
   acts_as_voter
+
+  def self.find_for_database_authentication(warden_conditions)
+    conditions = warden_conditions.dup
+    if login = conditions.delete(:login)
+      where(conditions.to_h).where(["lower(username) = :value OR lower(email) = :value", { :value => login.downcase }]).first
+    elsif conditions.has_key?(:username) || conditions.has_key?(:email)
+      conditions[:email].downcase! if conditions[:email]
+      where(conditions.to_h).first
+    end
+  end
+
+  before_save do
+    self.email.downcase! if self.email
+  end
+
+  def self.find_for_authentication(conditions)
+    conditions[:email].downcase!
```

Comment:
I just noticed that this also doesn't handle usernames. Can you try to figure out when, where, and why `find_for_database_authentication` and `find_for_authentication` are used? There's a lot of metaprogramming in the Devise library (to support arbitrary custom models like User, Admin, Customer, whatever you want to name your model), so it's going to be kind of difficult. But if you can figure out the reasoning for having the two finders, it would help a lot.